pyanaconda/core/subscription.py

```diff
@@ -38,3 +53,191 @@ def check_system_purpose_set(sysroot="/"):
+# thread. The simple scheduler hosts the code that starts the respective subscription
+# handling thread, which assures appropriate tasks are run.
+#
+# As the scheduler code than can be run either during early startup or in reaction to user
+# interaction in the Subscription spoke we avoid code duplication.
+
+
+def dummy_progress_callback(phase):
+    """Dummy progress reporting function used if no custom callback is set."""
+    pass
+
+
+def dummy_error_callback(error_message):
+    """Dummy error reporting function used if no custom callback is set."""
+    pass
+
+
+def register_and_subscribe(progress_callback=None, error_callback=None):
+    """Try to register and subscribe the installation environment.
+
+    :param progress_callback: progress callback function, takes one argument, subscription phase
+    :type progress_callback: a callable
+    :param error_callback: error callback function, takes one argument, the error message
+    :param error_callback: a callable
+    """
+
+    # assign dummy callback functions if none were provided by caller
+    if progress_callback is None:
+        progress_callback = dummy_progress_callback
+    if error_callback is None:
+        error_callback = dummy_error_callback
+
+    # connect to the Subscription DBus module
+    subscription_proxy = SUBSCRIPTION.get_proxy()
+
+    # First make sure network connectivity is available
+    # by waiting for the connectivity check thread
+    # to finish, in case it is running, usually early
+    # during Anaconda startup.
+    threadMgr.wait(THREAD_WAIT_FOR_CONNECTING_NM)
+    # Then check if we are not already registered.
+    #
+    # In some fairly bizarre cases it is apparently
+    # possible that registration & attach will succeed,
+    # but the attached subscription will be incomplete
+    # and/or invalid. These cases will be caught by
+    # the subscription token check and marked as failed
+    # by Anaconda.
+    #
+    # It is also possible that registration succeeds,
+    # but attach fails.
+    #
+    # To make recovery and another registration attempt
+    # possible, we need to first unregister the already
+    # registered system, as a registration attempt on
+    # an already registered system would fail.
+    if subscription_proxy.IsRegistered:
+        log.debug("subscription thread: system already registered, unregistering")
+        progress_callback(SubscriptionPhase.UNREGISTER)
+        task_path = subscription_proxy.UnregisterWithTask()
+        task_proxy = SUBSCRIPTION.get_proxy(task_path)
+        try:
+            task.sync_run_task(task_proxy)
+        except UnregistrationError as e:
+            log.debug("subscription thread: unregistration failed: %s", e)
+            # Failing to unregister the system is an unrecoverable error,
+            # so we end there.
+            error_callback(str(e))
+            return
+        log.debug("Subscription GUI: unregistration succeeded")
```

Comment:
I would abstract this to a private method.

Reply:
I tried that and it was not worth it - it's ~5 lines and you would have to pass it the subscription proxy and then detect if there was an error or not. The invocation & return checking have the same or more lines.